commonjs-extension-resolution-loader/loader.js

```diff
@@ -1,27 +1,34 @@
-import { existsSync } from 'fs';
-import { createRequire } from 'module';
+import { builtinModules } from 'node:module';
 import { dirname } from 'path';
-import { URL, fileURLToPath, pathToFileURL } from 'url';
+import { cwd } from 'process';
+import { fileURLToPath, pathToFileURL } from 'url';
+import { promisify } from 'util';
 
-const require = createRequire(import.meta.url);
-const baseURL = pathToFileURL(process.cwd() + '/').href;
+import resolveCallback from 'resolve/async.js';
 
-export function resolve(specifier, context, defaultResolve) {
+const resolveAsync = promisify(resolveCallback);
+
+const baseURL = pathToFileURL(cwd() + '/').href;
+
+
+export async function resolve(specifier, context, next) {
   const { parentURL = baseURL } = context;
 
-  // `require.resolve` works with paths, not URLs, so convert to and from
+  if (specifier.startsWith('node:') || builtinModules.includes(specifier)) {
+    return next(specifier, context);
+  }
+
+  // `resolveAsync` works with paths, not URLs
   if (specifier.startsWith('file://')) {
     specifier = fileURLToPath(specifier);
   }
-  const basePath = dirname(fileURLToPath(parentURL));
-  const resolvedPath = require.resolve(specifier, {paths: [basePath]});
+  const parentPath = fileURLToPath(parentURL);
 
-  if (existsSync(resolvedPath)) {
-    return {
-      url: pathToFileURL(resolvedPath).href
-    };
-  }
+  const resolution = await resolveAsync(specifier, {
+    basedir: dirname(parentPath),
+    extensions: ['.js', '.json', '.node'],
```

Comment:
I wonder if loaders should be able to expose their own hooks, for example to accept new extensions ... without that, a TS loader would have to implement the same thing but with a different extension, which would require to perform the same filesystem checks multiple times 🤔

Reply:
So ambient loaders weren’t enough, now we need loaders for our loaders . . .

Reply:
> accept new extensions

I think the way to achieve this is for loaders to optionally load config files. There’s nothing stopping me from adding some code to the top of `loader.js` to look for a possibly existing config file, and if so to `import` it, and read the extensions list from there.

One loader could communicate with the following ones by attaching something to `globalThis`, I think, since that should be shared within the loaders thread (?). So that’s another way to define/share configuration.

Reply:
This is the sort of problem that I'm referring to when I say that the "helper functions" need to compose like hooks.

Reply:
Yes, they'll share globalThis (as long as we keep ambient loaders and lay loaders in the same thread).

But `context` may be a better vehicle, and will is guaranteed to work regardless of threading.

Reply:
Yeah, the blocker right now for the helper functions is the API design. I took a stab at it in https://github.com/nodejs/loaders/pull/94 but I feel like that pass leaves a lot to be desired. A similar example is in https://github.com/browserify/resolve#resolveid-opts-cb: see the functions it accepts in the options to override its own lower-level operations.